design/control-data-plane-separation/design.md

```diff
@@ -23,377 +23,118 @@ architecture does not allow the control plane and data plane to scale independen
 
-The agent uses the `Upload` command to upload large payloads, such as the nginx configuration, to the control plane. For
-our use case, we only need the ability to send the nginx configuration from the control plane to the agent. We do not
-need the agent to upload its configuration to the control plane. Ideally, we would like to disable this functionality in
-the agent.
+#### Further Requirements and Implementation Details
 
-For the full commander service definition, see
-this [file](https://github.com/nginx/agent/blob/main/sdk/proto/command_svc.proto).
+- The control plane will need to run a gRPC server for the agent to connect to.
+- When an agent connects to the control plane, the payload _should_ contain the hostname (pod name) and nginx instanceID of the registering agent. This can be used to keep track of all agents/pods that are connected.
+  - We need to be able to link an agent connection with a subscription. These are two different gRPC calls. `Subscribe` is where we actually send an nginx config to an agent. We need to ensure that we are sending the correct nginx config to the correct agent. Ideally we use metadata from the agent connection (maybe hostname, maybe we need middleware to extract token/uuid from grpc connection headers) and store that in the context. The context is then used in the `Subscribe` call, where we can extract the agent identifier and send the proper nginx config for that identifier.
 
-#### Command Message
+  Process: agent `Connects` to NGF. We get its identifier and pod name, add the identifier(s) to a context cache, track that connection, and create a subscription channel for it. Agent then `Subscribes`. The context passed in allows us to use the identifier to grab the proper subscription channel and listen on it. This channel will receive a `ConfigApplyRequest` when we have a new nginx config to write.
 
-Command Messages have the following structure:
+- If a single nginx deployment is scaled, we should ensure that all instances for that deployment receive the same config (and aren't treated as "different").
+- Each graph that the control plane builds internally should be directly tied to an nginx deployment.
+- Whenever the control plane sees an nginx instance become Ready, we send its config to it (it doesn't matter if this is a new pod or a restarted pod).
+- If no nginx instances exist, control plane should not send any configuration.
+- The control plane should check if a connection exists first before sending the config (all replicas, even non-leaders, should be able to send the config because the agent may connect to a non-leader).
+  - Different pods within an nginx deployment may connect to a different NGF control plane if the NGF control plane has also been scaled. So each NGF control plane may have a different list of connected agents that they would be sending to. However, each NGF pod should have the exact same internal graph of nginx configuration.
```

Comment:
> However, each NGF pod should have the exact same internal graph of nginx configuration

Are we guaranteeing this to be true at any instance in time? Or are we just guaranteeing it will be eventually consistent?

Reply:
Eventually consistent in this case.